File: sdk/java/src/main/java/com/ouc/tcp/sdk/rdt1/Rdt1Sender.java
package com.ouc.tcp.sdk.rdt1;

import com.ouc.tcp.sdk.Packet;
import com.ouc.tcp.sdk.SystemContext;
import com.ouc.tcp.sdk.TcpHeader;
import com.ouc.tcp.sdk.TransportProtocol;

public final class Rdt1Sender implements TransportProtocol {
    @Override
    public void init(SystemContext ctx) {
        ctx.log("Java RDT1 sender ready");
    }

    @Override
    public void onPacket(SystemContext ctx, Packet packet) {
        ctx.log("RDT1 sender ignoring inbound packet seq=" + packet.getHeader().getSeqNum());
    }

    @Override
    public void onTimer(SystemContext ctx, int timerId) {
        ctx.log("RDT1 sender ignores timer " + timerId);
    }

    @Override
    public void onAppData(SystemContext ctx, byte[] data) {
        Packet packet = new Packet(new TcpHeader(), data);
        ctx.log("RDT1 sender sending " + data.length + " bytes");
        ctx.sendPacket(packet);
    }
}

File: sdk/cpp/src/rdt1_sender.cpp
#include "tcp_lab/sdk.hpp"

using namespace tcp_lab::sdk;

class Rdt1Sender final : public Protocol {
  public:
    void init() override {
        log("C++ RDT1 sender ready");
    }

    void on_app_data(const std::vector<uint8_t>& data) override {
        TcpHeader header{};
        std::vector<uint8_t> payload = data;
        log("RDT1 sender forwarding " + std::to_string(payload.size()) + " bytes");
        send_packet(header, payload);
    }
};

TCP_LAB_REGISTER_PROTOCOL(Rdt1Sender)
